File: app/javascript/channels/notificaciones_channel.js
import consumer from "./consumer"

let notificacionesSubscription = null;

// Funcin para suscribirse al canal de notificaciones
export function subscribeToNotificaciones(usuarioId, callbacks = {}) {
  // Cancelar cualquier suscripcin existente
  if (notificacionesSubscription) {
    notificacionesSubscription.unsubscribe();
  }
  
  // Crear nueva suscripcin
  notificacionesSubscription = consumer.subscriptions.create(
    { channel: "NotificacionesChannel", usuario_id: usuarioId },
    {
      connected() {
        console.log("Conectado al canal de notificaciones");
        if (callbacks.onConnected) callbacks.onConnected();
      },
      
      disconnected() {
        console.log("Desconectado del canal de notificaciones");
        if (callbacks.onDisconnected) callbacks.onDisconnected();
      },
      
      received(data) {
        console.log("Nueva notificacin recibida:", data);
        if (callbacks.onReceived) callbacks.onReceived(data);
      },
      
      // Mtodo para marcar una notificacin como leda
      marcarComoLeida(id) {
        this.perform('marcar_como_leida', { id });
      },
      
      // Mtodo para marcar todas las notificaciones como ledas
      marcarTodasComoLeidas() {
        this.perform('marcar_como_leida', { todas: true });
      }
    }
  );
  
  return notificacionesSubscription;
}

// Exportar tambin la suscripcin actual
export { notificacionesSubscription };

File: app/channels/notificaciones_channel.rb
# app/channels/notificaciones_channel.rb
class NotificacionesChannel < ApplicationCable::Channel
  def subscribed
    # Solo permitir suscripción a las notificaciones del usuario actual
    if current_usuario && params[:usuario_id] == current_usuario.id
      stream_from "notificaciones_#{current_usuario.id}"
    else
      reject
    end
  end

  def unsubscribed
    stop_all_streams
  end

  # Método llamado desde el cliente para marcar como leída
  def marcar_como_leida(data)
    return unless current_usuario

    if data['id'].present?
      # Marcar una notificación específica
      notificacion = current_usuario.notificaciones.find_by(id: data['id'])
      notificacion&.marcar_como_leida!
    elsif data['todas']
      # Marcar todas las notificaciones
      NotificacionesService.marcar_como_leidas(current_usuario)
    end
  end
end
